model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java

```diff
@@ -152,14 +176,32 @@ public long delete(QueryParameters<M> queryParameters) {
 
         CriteriaBuilder cb = em.getCriteriaBuilder();
 
+        // Remove all entities that are in the persistence context and that match the criteria.
+        // This avoids calling flush and clear which would detach all other unrelated entities as well.
+        int[] removed = {0};
+        MapModelCriteriaBuilder<String, E, M> mapMcb = queryParameters.getModelCriteriaBuilder().flashToModelCriteriaBuilder(createCriteriaBuilderMap());
+        cacheWithinSession.entrySet().removeIf(entry -> {
+            if (mapMcb.getKeyFilter().test(entry.getKey()) && mapMcb.getEntityFilter().test(entry.getValue())) {
+                em.remove(em.getReference(entityType, UUIDKey.INSTANCE.fromString(entry.getKey())));
```

Comment:
idea: does it make sense to "just" detach the entities (or evict()) which are about to be deleted and let the bulk delete handle the actual deletion? 

That way we'd mark entities in persistence context AND avoid having sql query per entity in the same time.

Reply:
I like the idea of activating the bulk delete in the future and keep the suggested change as is. Overall this whole thing gets a +1 from me.

Reply:
I've been giving it a thought that those instances that are created through the constructor of the entities that don't live a managed life within the entity manager could be skipped for the bulk delete iteration, still it is difficult for me to reason if they should be cleared from the cache or not; eventually they might "promote" their delegate to a managed entity ... for now I'd like to keep it as is and see if we hit a performance issue (what I don't think we will). 

So I'll put it on today's agenda, and hopefully get an approval either before or after the meeting to get it merged.

Reply:
If it's possible to activate bulk delete by hibernate's option (I was not aware of it and I was not able to find the option yet), then I agree with the approach.

One additional note: with this change we are modifying behavior for `read(String key)` method as prior this change json field was always loaded as it was expected that it'll be most likely needed. After this change we load whatever is defined within `JpaMapKeycloakTransaction.selectCbConstruct(CriteriaBuilder, Root<RE>)`.